crypto/bio/bss_dgram.c

```diff
@@ -808,6 +989,620 @@ static int dgram_puts(BIO *bp, const char *str)
+}
+#endif
+
+/*
+ * Converts flags passed to BIO_sendmmsg or BIO_recvmmsg to syscall flags. You
+ * should mask out any system flags returned by this function you cannot support
+ * in a particular circumstance. Currently no flags are defined.
+ */
+#if M_METHOD != M_METHOD_NONE
+static int translate_flags(uint64_t flags) {
+    return 0;
+}
+#endif
+
+static ossl_ssize_t dgram_sendmmsg(BIO *b, BIO_MSG *msg, size_t stride, size_t num_msg, uint64_t flags) {
+#if M_METHOD != M_METHOD_NONE
+    int ret;
+#endif
+#if M_METHOD == M_METHOD_RECVMMSG
+# define BIO_MAX_MSGS_PER_CALL   64
+    int sysflags;
+    bio_dgram_data *data = (bio_dgram_data *)b->ptr;
+    size_t i, j, jlim;
+    ossl_ssize_t num_done = 0;
+    struct mmsghdr mh[BIO_MAX_MSGS_PER_CALL];
+    struct iovec iov[BIO_MAX_MSGS_PER_CALL];
+    unsigned char control[BIO_CMSG_ALLOC_LEN][BIO_MAX_MSGS_PER_CALL];
+    int have_local_enabled = data->local_addr_enabled;
+#elif M_METHOD == M_METHOD_RECVMSG
+    int sysflags;
+    bio_dgram_data *data = (bio_dgram_data *)b->ptr;
+    size_t i;
+    struct msghdr mh;
+    struct iovec iov;
+    unsigned char control[BIO_CMSG_ALLOC_LEN];
+    int have_local_enabled = data->local_addr_enabled;
+#elif M_METHOD == M_METHOD_WSARECVMSG
+    bio_dgram_data *data = (bio_dgram_data *)b->ptr;
+    int have_local_enabled = data->local_addr_enabled;
+    WSAMSG wmsg;
+    WSABUF wbuf;
+    DWORD num_bytes_sent = 0;
+    unsigned char control[BIO_CMSG_ALLOC_LEN];
+#endif
+#if M_METHOD == M_METHOD_RECVFROM || M_METHOD == M_METHOD_WSARECVMSG
+    int sysflags;
+#endif
+
+    if (num_msg == 0)
+        return 0;
+
+#if M_METHOD != M_METHOD_NONE
+    sysflags = translate_flags(flags);
+#endif
+
+#if M_METHOD == M_METHOD_RECVMMSG
+    /*
+     * If MSG_WAITFORONE is defined, take this as a cue that recvmmsg is
+     * available. The messages passed to us are chunked into one or more
+     * recvmmsg calls. We have to do this because we need to allocate our
+     * translated struct msghdr and struct iovec on the stack to support
+     * multithreaded use.
+     */
+    for (i=0; i<num_msg; i += BIO_MAX_MSGS_PER_CALL) {
+        /* Translate all messages in this batch */
+        jlim = BIO_MAX_MSGS_PER_CALL;
+        if (jlim > num_msg-i)
+            jlim = num_msg-i;
+        for (j=0; j<jlim; ++j) {
+            translate_msg(&mh[j].msg_hdr, &iov[j], control[j], &BIO_MSG_N(msg, i+j));
+
+            /* If local address was requested, it must have been enabled */
+            if (BIO_MSG_N(msg, i+j).local != NULL) {
+                if (!have_local_enabled)
+                    /*
+                     * If we have done at least one message, we must return the
+                     * count; if we haven't done any, we can give an error code
+                     */
+                    return num_done > 0 ? num_done : -3;
+
+                if (pack_local(b, &mh[j].msg_hdr, BIO_MSG_N(msg, i+j).local) < 1)
+                    return num_done > 0 ? num_done : -3;
```

Comment:
Why the -3 special return code? If we really need to distinguish it, it should be a defined constant but I do not really see the need. I.E., why not -1?